# Data Science Jobs Scraper

In [147]:
import re
import requests
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup

### Import Target Keywords

In [148]:
df = pd.read_csv("ds_skills.csv")

df["Skill"] = df["Skill"].str.lower()
df["Category"] = df["Category"].str.lower()
df['total_jobs'] = 0

print("Total Skills: ", df.shape[0], " technologies")
df.head()

Total Skills:  50  technologies


Skill     Category  total_jobs
0      python  programming           0
1           r  programming           0
2         sql     database           0
3        java  programming           0
4  javascript  programming           0

In [149]:
skills = df['Skill'].to_list()

print(skills)

['python', 'r', 'sql', 'java', 'javascript', 'c', 'scala', 'julia', 'sas', 'matlab', 'matplotlib', 'seaborn', 'bokeh', 'plotly', 'ggplot', 'numpy', 'scipy', 'pandas', 'scikit learn', 'keras', 'pytorch', 'xgboost', 'tensorflow', 'aws', 'azure', 'haskel', 'swift', 'octave', 'perl', 'lisp', 'excel', 'power bi', 'tableau', 'google cloud', 'spark', 'hadoop', 'git', 'spss', 'stata', 'pig', 'hbase', 'docker', 'kubernetes', 'statsmodels', 'nltk', 'bitbucket', 'github', 'gitlab', 'nosql', 'hive']


## Indeed

In [150]:
url = f'https://www.indeed.com/jobs?q=%22data+scientist%22&l=United+States'

try:
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    count_str = soup.find('div', id="searchCountPages").get_text()
    numb = count_str.split()
    indeed_total = int(re.sub(",", "", numb[-2]))

except Exception as e:
    print(f'error: {e}')

print(indeed_total, "Data Science Jobs")

4758 Data Science Jobs


In [151]:
indeed_list = []

for skill in skills:
    url = f'https://www.indeed.com/jobs?q=%22data+scientist%22+%22{skill}%22&l=United+States'
    
    try:
        r = requests.get(url)
        soup = BeautifulSoup(r.text, 'html.parser')
        count_str = soup.find('div', id="searchCountPages").get_text()
        numb = count_str.split()
        indeed_list.append(numb[-2])
    except Exception as e:
        print(f'error: {e}')
        
print(indeed_list)

['3,621', '2,458', '2,623', '1,059', '273', '415', '640', '64', '720', '444', '128', '40', '84', '56', '87', '302', '151', '364', '415', '271', '403', '65', '678', '1,104', '490', '1', '10', '40', '105', '2', '494', '244', '1,069', '181', '1,230', '884', '255', '166', '59', '140', '104', '280', '188', '35', '44', '29', '173', '25', '327', '462']


In [152]:
indeed = []

for i in indeed_list:
    indeed.append(int(re.sub(",", "", i)))

df['total_jobs'] = df['total_jobs'] + indeed_total
df['indeed'] = indeed
df.head()

Skill     Category  total_jobs  indeed
0      python  programming        4758    3621
1           r  programming        4758    2458
2         sql     database        4758    2623
3        java  programming        4758    1059
4  javascript  programming        4758     273

## Simply Hired

In [153]:
url = f'https://www.simplyhired.com/search?q=%22data+scientist%22&l=United+States'
    
try:
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    count_str = soup.find('span', class_="CategoryPath-total").get_text()
    simply_total = int(re.sub(",", "", count_str))

except Exception as e:
    print(f'error: {e}')

print(simply_total, "Data Science Jobs")

3523 Data Science Jobs


In [154]:
simply_list = []

for skill in skills:
    url = f'https://www.simplyhired.com/search?q=%22data+scientist%22+%22{skill}%22&l=United+States'
    
    try:
        r = requests.get(url)
        soup = BeautifulSoup(r.text, 'html.parser')
        count_str = soup.find('span', class_="CategoryPath-total").get_text()
        simply_list.append(count_str)
    except Exception as e:
        print(f'error: {e}')
        
print(simply_list)

['2,710', '1,814', '1,995', '762', '177', '319', '453', '61', '534', '320', '94', '29', '46', '53', '72', '225', '120', '296', '285', '216', '304', '47', '526', '794', '397', '1', '9', '13', '90', '2', '405', '204', '753', '123', '931', '695', '218', '132', '54', '108', '87', '192', '155', '24', '41', '29', '136', '23', '270', '350']


In [155]:
simply = []

for i in simply_list:
    simply.append(int(re.sub(",", "", i)))

df['total_jobs'] = df['total_jobs'] + simply_total
df['simply'] = simply
df.head()

Skill     Category  total_jobs  indeed  simply
0      python  programming        8281    3621    2710
1           r  programming        8281    2458    1814
2         sql     database        8281    2623    1995
3        java  programming        8281    1059     762
4  javascript  programming        8281     273     177

## Glassdoor

In [156]:
url = f'https://www.glassdoor.com/Job/jobs.htm?sc.keyword="data+scientist"+"%22python%22"'

r = requests.get(url)
print(r.status_code)

403


**Status code is 403 (Forbidden)**. Therefore, we can not scrape content from Glassdoor. Optionally, we can try using **Selenium** in the future.

## Zip Recruiter

In [157]:
url = f'https://www.ziprecruiter.com/candidate/search?search=data+scientist&location=United+States'

r = requests.get(url)
print(r.status_code)

403


**Status code is 403 (Forbidden)**. Therefore, we can not scrape content from Glassdoor. Optionally, we can try using **Selenium** in the future.

## Career Builder

In [158]:
url = 'https://www.careerbuilder.com/jobs?keywords=Data+Scientist&location=United+States'

try:
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    count_str = soup.find('h1', class_="fz1rem").get_text()
    career_total = int(re.sub(",", "", count_str.split()[0]))

except Exception as e:
    print(f'error: {e}')

print(career_total, "Data Science Jobs")

3385 Data Science Jobs


In [159]:
career_list = []

for skill in skills:
    keyword = str(re.sub(",", "+", skill))
    url = f'https://www.careerbuilder.com/jobs?keywords={keyword}+Data+Scientist&location=United+States'
    
    try:
        r = requests.get(url)
        soup = BeautifulSoup(r.text, 'html.parser')
        count_str = soup.find('h1', class_="fz1rem").get_text()
        career_list.append(int(re.sub(",", "", count_str.split()[0])))
    
    except Exception as e:
        career_list.append(0)
        print(f'error: {e}')
        print(skill)
        
print(career_list)

error: 'NoneType' object has no attribute 'get_text'
haskel
[889, 1081, 694, 339, 104, 490, 152, 4, 168, 107, 30, 12, 6, 6, 25, 71, 31, 68, 54, 54, 48, 15, 103, 286, 169, 0, 1, 1, 53, 1, 577, 52, 291, 46, 298, 270, 110, 24, 11, 62, 37, 117, 74, 3, 14, 9, 47, 6, 107, 165]


In [160]:
career = []

for i in career_list:
    career.append(int(i))

df['total_jobs'] = df['total_jobs'] + career_total
df['career'] = career
df.head()

Skill     Category  total_jobs  indeed  simply  career
0      python  programming       11666    3621    2710     889
1           r  programming       11666    2458    1814    1081
2         sql     database       11666    2623    1995     694
3        java  programming       11666    1059     762     339
4  javascript  programming       11666     273     177     104

### Export Dataset

In [145]:
print(df.shape)
df.to_csv("ds_scraped_data.csv")

(50, 6)
